File: src/lib/api/feedback.ts
/**
 * AI Assignment Feedback API Client (Agent #8 Integrator)
 * =======================================================
 * Calls backend AI Assignment Feedback Agent endpoints to analyze code & written submissions.
 */

export interface AssignmentFeedbackResponse {
  assignment_title: string;
  subject: string;
  overall_score: number;
  letter_grade: string;
  error_identification: string[];
  explanation_of_mistakes: string;
  suggestions_for_improvement: string[];
  learning_resources: string[];
  refactored_solution_snippet: string;
  planner_recommendation: string;
}

import { apiFetch, API_BASE_URL } from "./client";

export async function generateAssignmentFeedback(
  studentId: string,
  assignmentTitle: string,
  submissionText: string,
  submissionType = "code",
  subject = "Computer Science / Mathematics"
): Promise<AssignmentFeedbackResponse> {
  const response = await apiFetch(`${API_BASE_URL}/ai/assignment/feedback`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      student_id: studentId || "demo_student",
      assignment_title: assignmentTitle,
      submission_text: submissionText,
      submission_type: submissionType,
      subject: subject,
    }),
  });

  if (!response.ok) {
    throw new Error(`Assignment Feedback failed with status ${response.status}`);
  }

  return await response.json();
}

File: reference-backend/app/api/routes/feedback.py
"""
AI Assignment Feedback Agent Routes (feedback.py)
=================================================

Exposes endpoints for analyzing student code submissions, mathematical derivations,
and written assignment solutions to generate structured 4-part AI feedback reports.
"""

import logging
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session as DBSession

from app.api.deps import get_optional_current_user
from app.db.session import get_db
from app.models.models import User
from app.schemas.schemas import (
    AssignmentFeedbackRequest,
    AssignmentFeedbackResponse,
)
from app.ai.agents.feedback_agent import feedback_agent

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/ai", tags=["Agent #8: AI Assignment Feedback Agent"])


@router.post("/assignment/feedback", response_model=AssignmentFeedbackResponse, status_code=status.HTTP_200_OK)
async def generate_assignment_feedback(
    req: AssignmentFeedbackRequest,
    db: DBSession = Depends(get_db),
    current_user: User | None = Depends(get_optional_current_user),
) -> AssignmentFeedbackResponse:
    """
    Analyzes student code/written submissions and returns structured 4-part AI feedback.
    """
    student_id = req.student_id or (current_user.id if current_user else "demo_student")

    try:
        res = await feedback_agent.analyze_submission(
            db=db,
            student_id=student_id,
            assignment_title=req.assignment_title,
            submission_text=req.submission_text,
            submission_type=req.submission_type,
            subject=req.subject,
        )
        return AssignmentFeedbackResponse(**res)
    except Exception as e:
        logger.error("Failed to generate assignment feedback: %s", str(e), exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"AI Assignment Feedback Agent error: {str(e)}",
        )
